File: app/src/main/java/com/mrboomdev/awery/catalog/extensions/support/tachiyomi/TachiyomiManager.java
package com.mrboomdev.awery.catalog.extensions.support.tachiyomi;

import android.util.Log;

import com.mrboomdev.awery.catalog.extensions.Extension;
import com.mrboomdev.awery.catalog.extensions.support.yomi.YomiManager;

import eu.kanade.tachiyomi.source.MangaSource;
import eu.kanade.tachiyomi.source.SourceFactory;

public class TachiyomiManager extends YomiManager {

	@Override
	public String getName() {
		return "Tachiyomi";
	}

	@Override
	public String getId() {
		return "TACHIYOMI_KOTLIN";
	}

	@Override
	public String getMainClassMeta() {
		return "tachiyomi.extension.class";
	}

	@Override
	public String getNsfwMeta() {
		return "tachiyomi.extension.nsfw";
	}

	@Override
	public String getRequiredFeature() {
		return "tachiyomi.extension";
	}

	@Override
	public String getPrefix() {
		return "Tachiyomi: ";
	}

	@Override
	public double getMinVersion() {
		return 1.2;
	}

	@Override
	public double getMaxVersion() {
		return 1.5;
	}

	@Override
	public void fillProvider(Extension extension, Object main) {
		if(main instanceof MangaSource source) {
			extension.addProvider(new TachiyomiProvider(source));
		} else if(main instanceof SourceFactory factory) {
			for(var source : factory.createSources()) {
				fillProvider(extension, source);
			}
		} else {
			Log.e("TachiyomiManager", "Unknown source type: " + main.getClass().getName());
		}
	}
}

File: app/src/main/java/com/mrboomdev/awery/sources/yomi/tachiyomi/TachiyomiManager.kt
package com.mrboomdev.awery.sources.yomi.tachiyomi

import android.content.Context
import android.content.pm.PackageInfo
import com.mrboomdev.awery.ext.constants.AweryAgeRating
import com.mrboomdev.awery.sources.yomi.YomiManager
import com.mrboomdev.awery.util.AndroidImage
import eu.kanade.tachiyomi.animesource.AnimeSource
import eu.kanade.tachiyomi.animesource.AnimeSourceFactory
import eu.kanade.tachiyomi.source.MangaSource
import eu.kanade.tachiyomi.source.SourceFactory

class TachiyomiManager(
	context: Context
): YomiManager<MangaSource, TachiyomiSource>(context) {
	override val minVersion = 1.2
	override val maxVersion = 1.5
	override val name = "Tachiyomi"
	override val id = ID

	override val appLabelPrefix = "Tachiyomi: "
	override val nsfwMeta = "tachiyomi.extension.nsfw"
	override val mainClass = "tachiyomi.extension.class"
	override val requiredFeature = "tachiyomi.extension"

	private fun getSelected(it: Any?): MangaSource? {
		if(it == null) {
			return null
		}

		if(it is MangaSource) {
			return it
		}

		if(it is SourceFactory) {
			return getSelected(it.createSources())
		}

		if(it is List<*>) {
			return getSelected(it.toTypedArray())
		}

		if(it is Array<*>) {
			for(i in it) {
				if(i is SourceFactory) {
					return getSelected(i.createSources())
				}

				if(i is MangaSource) {
					// TODO: Return an source selected by the user.
					return i
				}
			}
		}

		throw UnsupportedOperationException("Unsupported source type!")
	}

	@Throws(IllegalArgumentException::class)
	override fun createSourceWrapper(
		label: String,
		isNsfw: Boolean,
		packageInfo: PackageInfo,
		sources: Array<Any>?,
		exception: Throwable?
	): TachiyomiSource {
		val selectedSource = getSelected(sources)

		return object : TachiyomiSource(packageInfo, selectedSource) {
			override val manager = this@TachiyomiManager
			override val exception = exception

			override val ageRating = (if(isNsfw) {
				AweryAgeRating.NSFW
			} else AweryAgeRating.EVERYONE).toString()

			override val icon = AndroidImage(packageInfo
				.applicationInfo!!.loadIcon(context.packageManager))

			override val name by lazy {
				selectedSource?.name ?: label
			}
		}
	}

	override fun getSourceLongId(source: MangaSource): Long {
		return source.id
	}

	companion object {
		const val ID = "TACHIYOMI_KOTLIN"
	}
}